### PR TITLE
Add api tests for sharing with expiration date

### LINKED ISSUE #36545: Test: the fields of the last response should not include does not work with 'expiration' field
The following step does not work if we try to check for `ANY_VALUE`:
```feature
Then the fields of the last response should not include
  | expiration | ANY_VALUE |
```

### Analysis:
[This is due to how `strtotime()` works](https://www.php.net/manual/en/function.strtotime.php#refsect1-function.strtotime-returnvalues). This needs to be handled somewhere here:
https://github.com/owncloud/core/blob/f37a6dc2774d54c16b3fe9b9cff30138bc4a23d7/tests/acceptance/features/bootstrap/Sharing.php#L662-L671

strtotime is passed `ANY_VALUE` which when fails returns `FALSE`. That, when passed to `date()` 
returns the Unix epoch (i.e. 1st January 1970), and the content to assert is no longer `ANY_VALUE`.

This when passed to `doesFieldValueMatchExpectedContent()` then returns false even though the field is still there.

https://github.com/owncloud/core/blob/f37a6dc2774d54c16b3fe9b9cff30138bc4a23d7/tests/acceptance/features/bootstrap/Sharing.php#L629-L645

### PR DESCRIPTION
**Pull request by skshetry**
## Description

This adds some API tests for sharing with expiration date.

## Related Issue

Related to owncloud/enterprise#2001. Fixes #36545 as well.

## How Has This Been Tested?

- locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 1]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | permissions | read,share     |
+      | expireDate  | +15 days       |
+    And the administrator sets parameter "shareapi_default_expire_date_user_share" of app "core" to "no"
+    And user "user0" gets the info of the last share using the sharing API
+    Then the fields of the last response should include
+      | share_type  | user                 |
+      | file_target | /FOLDER              |
+      | uid_owner   | user0                |
+      | share_with  | user1                |
+      | expiration  | +15 days             |
```
[the remark this answers is not in the document]
Comment (phil-davis) [reply tests/acceptance/features/apiShareManagementBasic/createShare.feature]: I asked in #36550

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 2]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | ocs_api_version |
+      | 1               |
+      | 2               |
+
+  Scenario Outline: sharing with expiration date is enabled and enforced, user shares with expiration date and then disables
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 3]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+    Then the OCS status code should be "<ocs_status_code>"
+    And the HTTP status code should be "<http_status_code>"
+    And the fields of the last response should include
+      | expiration | |
+    Examples:
+      | ocs_api_version | ocs_status_code | http_status_code |
+      | 1               | 100             | 200              |
+      | 2               | 200             | 200              |
+
+  Scenario Outline: sharing with expiration date enabled but not enforced for groups, users shares with expiration date
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
  Scenario Outline: sharing with expiration date enabled but not enforced for groups, user shares with expiration date
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 4]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 5]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | ocs_api_version | ocs_status_code | http_status_code |
+      | 1               | 100             | 200              |
+      | 2               | 200             | 200              |
+
+  Scenario Outline: sharing with expiration date is not enabled for groups, user shares with expiration date set
+    Given using OCS API version "<ocs_api_version>"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 6]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 1               | 100             | 200              |
+      | 2               | 200             | 200              |
+
+  Scenario Outline: sharing with expiration date is enabled but not enforced for groups, user shares with expiration date and then disables
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 7]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 2               |
+
+  Scenario Outline: sharing with expiration date is enabled and enforced for groups, user shares with expiration date and then disables
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 8]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | ocs_api_version |
+      | 1               |
+      | 2               |
+
+  Scenario Outline: sharing with expiration date is enabled and enforced for users, user shares with expiration date more than the default
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 9]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 1               |
+      | 2               |
+
+  Scenario Outline: sharing with expiration date is enabled and enforced for users/max expire date set, user shares with expiration date more than the max expire date
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "30"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 10]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 2               |
+
+  Scenario Outline: sharing with expiration date is enabled and enforced for groups, user shares with expiration date more than the default
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 11]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+
+  Scenario Outline: sharing with expiration date is enabled and enforced for users/max expire date set for groups, user shares with expiration date more than the max expire date
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_group_share" of app "core" has been set to "yes"
+    And parameter "shareapi_expire_after_n_days_group_share" of app "core" has been set to "30"
+    And user "user1" has been created with default attributes and without skeleton files
+    And group "grp1" has been created
+    And user "user1" has been added to group "grp1"
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 12]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | ocs_api_version |
+      | 1               |
+      | 2               |
+
+  Scenario Outline: sharing with expiration date enforced for users, user shares with invalid expiration date set
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 13]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | ocs_api_version | ocs_status_code | http_status_code |
+      | 1               | 404             | 200              |
+      | 2               | 404             | 404              |
+
+  Scenario Outline: sharing with expiration date enforced for users, user shares with past expiration date set
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 14]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 1               | 404             | 200              |
+      | 2               | 404             | 404              |
+
+  Scenario Outline: sharing with expiration date set, max expire date is 0, user shares without specifying expiration date
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "0"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/acceptance/features/apiShareManagementBasic/createShare.feature [hunk 15]
```diff
@@ -1596,3 +1596,554 @@ Feature: sharing
...
+      | 1               |
+      | 2               |
+
+  Scenario Outline: sharing with expiration date set, max expire date is 1, user shares without specifying expiration date
+    Given using OCS API version "<ocs_api_version>"
+    And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_enforce_expire_date_user_share" of app "core" has been set to "yes"
+    And parameter "shareapi_expire_after_n_days_user_share" of app "core" has been set to "1"
+    And user "user1" has been created with default attributes and without skeleton files
+    When user "user0" has created a share with settings
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareManagementBasic/createShare.feature]:
```suggestion
    When user "user0" creates a share using the sharing API with settings
```

File: tests/TestHelpers/SharingHelper.php [hunk 16]
```diff
@@ -140,7 +140,13 @@ public static function createShare(
 			$fd['name'] = $linkName;
 		}
 		if ($expireDate !== null) {
-			$fd['expireDate'] = \date('Y-m-d', \strtotime($expireDate));
+			$timestamp = \strtotime($expireDate);
+			if ($timestamp === false) {
+				$fd['expireDate'] = $expireDate;
+				print("expireDate (value: '$expireDate') could not be parsed into time, sending as it is.");
+			} else {
+				$fd['expireDate'] = \date('Y-m-d', \strtotime($expireDate));
```
Comment (skshetry) [inline tests/TestHelpers/SharingHelper.php]:
```suggestion
				$fd['expireDate'] = \date('Y-m-d', $timestamp);
```

File: tests/acceptance/features/bootstrap/Sharing.php [hunk 17]
```diff
@@ -1201,7 +1208,7 @@ public function userGetsInfoOfLastShareUsingTheSharingApi($user) {
 	 * @throws Exception
 	 */
 	public function getLastShareIdOf($user) {
-		if ($this->lastShareData !== null) {
+		if (isset($this->lastShareData->data[0]->id)) {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/Sharing.php]: I have changed this behavior slightly. Previously, even if the sharing failed, it'd return 0 (due to casting of not existing element in the xml) and would try to get the share of `/shares/0`.

So, the following scenario was testing this wrong behavior, and fails currently:
https://github.com/owncloud/core/blob/a3ab953eb8f2eec174cd2463eb332f0f1d1d1ebe/tests/acceptance/features/apiShareUpdate/updateShare.feature#L548-L556

We can just remove the last two lines.

File: tests/acceptance/features/apiShareUpdate/updateShare.feature [hunk 18]
```diff
@@ -552,8 +552,6 @@ Feature: sharing
     When user "user0" shares file "/welcome.txt" with group "grp1" using the sharing API
     Then the OCS status code should be "404"
     And the HTTP status code should be "<http_status_code>"
-    When user "user0" gets the info of the last share using the sharing API
-    Then the last response should be empty
```
Comment (skshetry) [inline tests/acceptance/features/apiShareUpdate/updateShare.feature]: This test is just wrong. There is no last share to fetch the info, as the sharing failed.

**Review by phil-davis — APPROVED** [review]

**Review by phil-davis — APPROVED** [review]

**Pull request merged**